### PR TITLE
Extend the list of default language extensions

### PR DESCRIPTION
**Pull request by ChrisPenner**
Chatted with @fisx this morning and came up with the following list of default extensions; feel free to chime in if there's anything you'd like to be added/removed, or just have thoughts about it 😄

Note that I'm leaving off `StrictData` for now since that may have unintended effects; we can look at that one separately. 

- AllowAmbiguousTypes        
- BangPatterns               
- ConstraintKinds            
- DataKinds                  
- DefaultSignatures          
- DeriveFunctor              
- DeriveGeneric              
- DeriveLift                 
- DeriveTraversable          
- EmptyCase                  
- FlexibleContexts           
- FlexibleInstances          
- FunctionalDependencies     
- GADTs                      
- InstanceSigs               
- KindSignatures             
- LambdaCase                 
- MultiParamTypeClasses      
- MultiWayIf                 
- NamedFieldPuns             
- NoImplicitPrelude          
- OverloadedStrings          
- PackageImports             
- PatternSynonyms            
- PolyKinds                  
- QuasiQuotes                
- RankNTypes                 
- ScopedTypeVariables        
- StandaloneDeriving         
- StandaloneDeriving         
- TemplateHaskell            
- TupleSections              
- TypeApplications           
- TypeFamilies               
- TypeFamilyDependencies     
- TypeOperators              
- UndecidableInstances       
- ViewPatterns               

Notably missing from the list are:
- GeneralizedNewtypeDeriving
    - @fisx suggested this not be globally enabled; presumably due to the following:
    - > GeneralizedNewtypeDeriving — It can be used to violate constructor access control, by allowing untrusted code to manipulate protected data types in ways the data type author did not intend. For example can be used to break invariants of data structures.
    - https://downloads.haskell.org/~ghc/7.2.1/docs/html/users_guide/safe-haskell.html
- RecordWildCards
    - I'm personally not a big fan of using RWC; at least not in a pattern-matching position as it brings into scope many names without any way to tell WHAT was brought into scope; this makes it tough to read, and also means that adding new record fields can cause some weird name shadowing problems in distant code blocks.

Ones which I've added but may be worth discussing:
- AllowAmbiguousTypes
    - Are there any cases where one could accidentally introduce an ambiguous type without noticing? I don't believe so, but correct me if I'm wrong.
- UndecidableInstances
    - Sometimes it's nice to know when you've written an instance which is undecidable (via getting a compiler warning), but in general I don't really care 🤷‍♂️ 

Changes in this PR other than enabling the modules are related to fixing issues caused by non-generalized let bindings due to `GADTs` enabling `MonoLocalBinds`. It's strange that this causes several problems for us, but in general I think it just requires that we be more explicit, which is a good thing. Since we do need GADTs in SOME parts of our code, I'm in favour of enabling it everywhere for consistency of the language; as expressed better by Alexis King here:

https://lexi-lambda.github.io/blog/2018/02/10/an-opinionated-guide-to-haskell-in-2018/

> All of these extensions involve a tradeoff. Enabling GADTs also implies MonoLocalBinds, which disables let generalization, one of the most likely ways a program that used to typecheck might subsequently fail to do so. Some might argue that this is a good reason to turn GADTs on in a per-module basis, but I disagree: I actually want my language to be fairly consistent, and given that I know I am likely going to want to use GADTs somewhere, I want MonoLocalBinds enabled everywhere, not inconsistently and sporadically.

### CODE REVIEW

File: services/gundeck/src/Gundeck/Push.hs [hunk 1]
```diff
@@ -344,7 +344,7 @@ addToken (uid ::: cid ::: req ::: _) = do
     continue t Nothing  = create (0 :: Int) t
     continue t (Just a) = update (0 :: Int) t (a^.addrEndpoint)
 
-    create :: Int -> PushToken -> Gundeck (Either Response (Address a))
+    create :: Int -> PushToken -> Gundeck (Either Response (Address "no-keys"))
```
Comment (ChrisPenner) [inline services/gundeck/src/Gundeck/Push.hs]: This fix was a little strange, I'm mostly confused as to why it worked before 🤔

File: services/spar/test-integration/Test/Spar/DataSpec.hs [hunk 2]
```diff
@@ -195,7 +195,7 @@ spec = do
 
 
 testSPStoreID
-  :: forall m a. (m ~ ReaderT Data.Env (ExceptT TTLError Client), Typeable a)
+  :: forall m (a :: *). (m ~ ReaderT Data.Env (ExceptT TTLError Client), Typeable a)
```
Comment (ChrisPenner) [inline services/spar/test-integration/Test/Spar/DataSpec.hs]: Now that PolyKinds is enabled this function is SO polymorphic it couldn't even tell which kind it wants; I just went with *.

File: services/spar/test-integration/Test/Spar/DataSpec.hs [hunk 3]
```diff
@@ -195,7 +196,7 @@ spec = do
 
 
 testSPStoreID
-  :: forall m a. (m ~ ReaderT Data.Env (ExceptT TTLError Client), Typeable a)
+  :: forall m (a :: Type). (m ~ ReaderT Data.Env (ExceptT TTLError Client), Typeable a)
```
Comment (ChrisPenner) [inline services/spar/test-integration/Test/Spar/DataSpec.hs]: Now that PolyKinds is enabled this function is SO polymorphic it couldn't even tell which kind it wants.

**Comment by fisx** [comment]
I like this.  I would even like it more if we could figure out how to maintain this list in only one place, not in the docs *and* in *every* cabal file.  If we used hpack for all packages, we might be able to have the list somewhere and import it into all package.yaml files.  And since that import would be yaml, it may even be possible to import it into the markdown docs.

Changing from manually maintained cabal files to hpack obviously would make this PR a lot bigger, so I'm also fine with doing this another time.

Also, would it make sense to add some stuff from the PR description to the docs as well, so it'll stick around?

Will look at the details soon!

File: doc/Style.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+- NoMonomorphismRestriction
+- OverloadedStrings
+- PackageImports
+- PatternSynonyms
+- PolyKinds
+- QuasiQuotes
+- RankNTypes
+- ScopedTypeVariables
+- StandaloneDeriving
+- StandaloneDeriving
```
Comment (neongreen) [inline doc/Style.md]: Duplicate.

**Comment by neongreen** [comment]
I'm okay with the list.

We still need to remove those extensions from actual files though.

**Comment by ChrisPenner** [comment]
Just realized that I missed a large amount of modules; there's a few more GADT related fixes so I'll do that tomorrow, or give up and just not add GADTs as a default 🤷‍♂️ 

Also I removed all the relevant pragmas using a handy :

```
for f in $(fd -e hs); do
    awk '!/LANGUAGE (AllowAmbiguousTypes|BangPatterns|ConstraintKinds|DataKinds|DefaultSignatures|DeriveFunctor|DeriveGeneric|DeriveLift|DeriveTraversable|EmptyCase|FlexibleContexts|FlexibleInstances|FunctionalDependencies|GADTs|InstanceSigs|KindSignatures|LambdaCase|MultiParamTypeClasses|MultiWayIf|NamedFieldPuns|NoImplicitPrelude|NoMonomorphismRestriction|OverloadedStrings|PackageImports|PatternSynonyms|PolyKinds|QuasiQuotes|RankNTypes|ScopedTypeVariables|StandaloneDeriving|StandaloneDeriving|TemplateHaskell|TupleSections|TypeApplications|TypeFamilies|TypeFamilyDependencies|TypeOperators|UndecidableInstances|ViewPatterns)/' "$f" | sponge "$f"
done;
```

So that'll be pushed once I get the GADTs stuff working; then we can put this to bed.

File: services/spar/package.yaml [hunk 5]
```diff
@@ -90,6 +128,8 @@ tests:
       - spar
       - uri-bytestring
 
+    default-extensions: *extension-list
```
Comment (neongreen) [inline services/spar/package.yaml]: I think you don't need that, just specifying default-extensions at the top level is enough (same as with `dependencies`).

**Comment by neongreen** [comment]
> Remove NoMonomorphismRestriction

Oh right, I didn't notice `NoMonomorphismRestriction` slipping in :/

**Pull request merged**